### PR TITLE
Use chunked upload for all put object.

### LINKED ISSUE #568: Add streaming signature support for putobject API.
- no md5sum. #567 
- only sha checksum for http protocol.
- all put objects have to use streaming signature v4 by default
- Use 8 threads by default optionally configurable.

### PR DESCRIPTION
**Pull request by balamurugana**
For more information
http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html

Fixes #568

### CODE REVIEW

**Comment by balamurugana** [comment]
encrypted putObject() fails with unknown reason.  Debugging it.

**Comment by balamurugana** [comment]
All issues are fixed.
All functional tests run successfully on play and Amazon S3.

**Comment by NitishT** [comment]
Could you please fix the conflict in `FunctionalTest.java` @balamurugana

**Comment by balamurugana** [comment]
travis-ci fails due to its internal problem.

**Comment by harshavardhana** [comment]
> travis-ci fails due to its internal problem.

oraclejdk8 failed for functional tests. 

```
The command "./gradlew check" exited with 0.
104.52s$ if java -version 2>&1 | grep -q 'java version "1.8.'; then ./gradlew runFunctionalTest; else true; fi
:api:compileJava UP-TO-DATE
:api:processResources UP-TO-DATE
:api:classes UP-TO-DATE
:api:jar UP-TO-DATE
:functional:compileJava
:functional:processResources UP-TO-DATE
:functional:classes
:functional:runFunctionalTest
Test: makeBucket(String bucketName)
Test: listBuckets()
Test: bucketExists(String bucketName)
Test: removeBucket(String bucketName)
Test: putObject(String bucketName, String objectName, String filename)
Test: multipart: putObject(String bucketName, String objectName, String filename)
Test: putObject(String bucketName, String objectName, String filename, String contentType)
Test: putObject(String bucketName, String objectName, InputStream body, long size, String contentType)
Test: multipart resume: putObject(String bucketName, String objectName, InputStream body, long size, String contentType)
Test: objectName with path segments: putObject(String bucketName, String objectName, InputStream body, long size, String contentType)
Test: putObject(String bucketName, String objectName, InputStream body, String contentType)
Test: multipart: putObject(String bucketName, String objectName, InputStream body, String contentType)
Test: putObject(String bucketName, String objectName, InputStream stream, long size, String contentType, SecretKey key).
Test: putObject(String bucketName, String objectName, InputStream stream, long size, String contentType, KeyPair keyPair).
Test: putObject(String bucketName, String objectName, InputStream stream, long size, Map<String, String> headerMap).
Test: statObject(String bucketName, String objectName)
Test: getObject(String bucketName, String objectName)
Test: getObject(String bucketName, String objectName, long offset)
Test: getObject(String bucketName, String objectName, long offset, Long length)
Test: getObject(String bucketName, String objectName, String filename)
Test: objectName with multiple path segments: getObject(String bucketName, String objectName, String filename)
Test: getObject(String bucketName, String objectName, SecretKey key).
:functional:runFunctionalTest FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':functional:runFunctionalTest'.
> Process 'command '/usr/lib/jvm/java-8-oracle/bin/java'' finished with non-zero exit value 137
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
BUILD FAILED
Total time: 1 mins 43.336 secs
The command "if java -version 2>&1 | grep -q 'java version "1.8.'; then ./gradlew runFunctionalTest; else true; fi" exited with 1.
```

**Comment by balamurugana** [comment]
@harshavardhana you would need to check other runs as well.

**Comment by harshavardhana** [comment]
> @harshavardhana you would need to check other runs as well.

Travis has updated their image to trusty which doesn't ship oraclejdk7 anymore - https://github.com/travis-ci/travis-ci/issues/7884  - Fix is to remove oraclejdk7

http://www.webupd8.org/2017/06/why-oracle-java-7-and-6-installers-no.html - We need to rework our .travis.yml

**Comment by NitishT** [comment]
Tested locally against AWS S3, tests pass.

File: api/src/main/java/io/minio/ChunkedInputStream.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        } else {
+          // Send final additional chunk to complete chunk upload.
+          byte[] chunk = new byte[0];
+          createChunkBody(chunk);
+        }
+      }
+
+      this.bytesRead++;
+      // Return value must be between 0 to 255.
+      return this.chunkBody[this.chunkPos++] & 0xFF;
```
Comment (krisis) [inline api/src/main/java/io/minio/ChunkedInputStream.java]: How does `this.chunkBody[this.chunkPos++]` evaluate? Does it return the value at index `this.chunkPos` before the increment? Is it non-idiomatic to do the increment and indexing in separate statements?

Comment (balamurugana) [inline api/src/main/java/io/minio/ChunkedInputStream.java]: > Does it return the value at index this.chunkPos before the increment?

Yes.  As `++` operator placed after variable, the variable is incremented after reading the value.  I am breaking the statement into two for better readability

**Comment by balamurugana** [comment]
Depends on https://github.com/minio/minio-java/pull/596 for travis-ci failure.

**Comment by balamurugana** [comment]
I see oraclejdk8 fails in travis-ci with below error
```
Test: getObject(String bucketName, String objectName, SecretKey key).

:functional:runFunctionalTest FAILED

FAILURE: Build failed with an exception.

* What went wrong:

Execution failed for task ':functional:runFunctionalTest'.

> Process 'command '/usr/lib/jvm/java-8-oracle/bin/java'' finished with non-zero exit value 137
```

here, exit code denotes signal 9 killed the process (formula: 128 + signal = 137).  It means OOM in linux.  Not sure why OOM only on oraclejdk8.

**Comment by harshavardhana** [comment]
> I see oraclejdk8 fails in travis-ci with below error
> 
> Test: getObject(String bucketName, String objectName, SecretKey key).
> 
> :functional:runFunctionalTest FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> 
> Execution failed for task ':functional:runFunctionalTest'.
> 
> > Process 'command '/usr/lib/jvm/java-8-oracle/bin/java'' finished

We might have to test it with oraclejdk8 . I don't have oraclejdk8 locally will install and check.

**Comment by balamurugana** [comment]
Travis CI succeeded without any change in the PR

**Comment by balamurugana** [comment]
@krisis @NitishT please add +1 which disappeared by my refresh

**Pull request merged**